### PR TITLE
A job template to remove failed disk

### PR DESCRIPTION
**Pull request by xenolinux**
This job template will manage to avoid the usage of the toolbox in order to use complex ceph commands for removing failed osd. A script can be executed with Ceph commands as if running in the toolbox, but will instead run in a job where the logs can be collected and analyzed separately.

### CODE REVIEW

File: pkg/controller/storagecluster/reconcile.go [hunk 1]
```diff
@@ -1037,3 +1038,96 @@ func getMonCount(nodeCount int) int {
...
+							VolumeMounts: []corev1.VolumeMount{
+								corev1.VolumeMount{
+									MountPath: "/etc/ceph",
+									Name:      "ceph-config",
+									ReadOnly:  true,
+								},
+							},
+							Command: []string{"/bin/bash", "-c"},
+							Args: []string{
+								"ceph osd out osd.FAILED_OSD_ID",   //this command marks osd out
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1092]: Does it work with the two commands as separate arguments in the string slice? In my testing the commands were a single multi-line parameter as seen [here](https://github.com/rook/rook/blob/master/cluster/examples/kubernetes/ceph/toolbox-job.yaml#L38-L43), but maybe this way works fine too.

File: must-gather/collection-scripts/gather [hunk 2]
```diff
@@ -32,19 +32,20 @@ mkdir -p ${BASE_COLLECTION_PATH}/oc_output/
 commands=()
 
 # collect oc output of OC commands
-commands+=("version")
```
Comment (travisn) [inline must-gather/collection-scripts/gather:35]: These changes aren't related to the new template, right? Seems like we should revert these changes.

File: pkg/controller/storagecluster/reconcile.go [hunk 3]
```diff
@@ -1037,3 +1038,96 @@ func getMonCount(nodeCount int) int {
 
 	return count
 }
+
+func newCleanupJob(sc *ocsv1.StorageCluster, RookImage string, OsdID int, reqLogger logr.Logger) *batchv1.Job {
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1042]: Remaining work items:
- Wrap the Job in an OpenShift template.
- Call this method from the reconcile loop so it is kept up-to-date. 
- When there is an upgrade we need to ensure the template is updated. For example, in the 4.5 release this template will likely change once we have a simplified rook job available.

File: must-gather/collection-scripts/gather (hunk 2 above)
Comment (obnoxxx) [reply must-gather/collection-scripts/gather:35]: Indeed, there seems to be a wird openshift-bot merge commit https://github.com/openshift/ocs-operator/pull/481/commits/5bb89d20ad5a8092582adb8ebea182fca8ad691d called "aa" from Mar 05 in this PR.

File: pkg/controller/storagecluster/reconcile.go (hunk 1 above)
Comment (ashishranjan738) [reply pkg/controller/storagecluster/reconcile.go]: I think we changed it to string slice.

File: pkg/controller/storagecluster/reconcile.go [hunk 4]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
 
 	return count
 }
+
+//ensureJobTemplates returns spec of a job template
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:1160]: I see it only returns `error`??

File: pkg/controller/storagecluster/reconcile.go [hunk 5]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
...
+func (r *ReconcileStorageCluster) ensureJobTemplates(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:1161]: This func looks way too complicated than it needs to be.

File: pkg/controller/storagecluster/reconcile.go [hunk 6]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
...
+	osdCleanUpTemplate := &openshiftv1.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sc.Name + "-osd-removal",
+			Namespace: sc.Namespace,
+		},
+	}
+	_, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, osdCleanUpTemplate, func() error {
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:1168]: Can we use `r.client.Create(context.TODO(), types.NamespacedName{...}, osdCleanUpTemplate)` for simplicity?

File: pkg/controller/storagecluster/reconcile.go [hunk 7]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
...
+		osdCleanUpTemplate.Parameters = []openshiftv1.Parameter{
+			{
+				Name:     "ROOK_IMAGE",
+				Required: true,
+			},
+			{
+				Name:     "FAILED_OSD_ID",
+				Required: true,
+			},
+		}
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:1183]: Can we move this to part where you initialize `osdCleanUpTemplate` ?
This doesn't look like it needs to be separate.

File: pkg/controller/storagecluster/reconcile.go [hunk 8]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
...
+		return controllerutil.SetControllerReference(sc, osdCleanUpTemplate, r.scheme)
+	})
+	if err != nil && !errors.IsAlreadyExists(err) {
+		return err
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:1187]: Add more context to the error message. Saves our debugging time later on.

File: pkg/controller/storagecluster/reconcile.go [hunk 9]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
...
+	}
+	return nil
+
```
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:1190]: remove empty line.

File: pkg/controller/storagecluster/reconcile.go (hunk 4 above)
Comment (umangachapagain) [inline pkg/controller/storagecluster/reconcile.go:1160]:
```suggestion
// ensureJobTemplates returns spec of a job template
```

File: cmd/manager/main.go [hunk 10]
```diff
@@ -120,6 +122,11 @@ func main() {
 		os.Exit(1)
 	}
 
+	if err := openshiftv1.AddToScheme(mgrScheme); err != nil {
+		log.Error(err, "Failed adding openshift/v1 to scheme")
+		os.Exit(1)
+	}
```
Comment (umangachapagain) [inline cmd/manager/main.go:128]: Why is this required?
I see it is also being used in multiple tests.

I never understood the purpose of this.

Comment (jarrpa) [reply cmd/manager/main.go:128]: If I remember correctly, this is because we need to be able to understand and manipulate the specs for all the API objects. Since Templates are an OpenShift construct, we need to add the entire openshift scheme.

File: pkg/controller/storagecluster/reconcile.go (hunk 4 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:1160]: That is the ensureFunc template, so yes that's all i can return.

File: pkg/controller/storagecluster/reconcile.go (hunk 8 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:1187]: Specifically, something like `failed to create Template: %v` where you pass `err` as the variable.

File: pkg/controller/storagecluster/reconcile.go (hunk 6 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:1168]: We should do proper reconciliation and make sure we either Create or Update as appropriate. This seems like an appropriate use of this helper function, and we should probably consider using it elsewhere... unless there's a reason not to? The alternative is certainly more complicated: https://github.com/openshift/ocs-operator/blob/master/pkg/controller/storagecluster/reconcile.go#L727-L763

File: pkg/controller/storagecluster/reconcile.go (hunk 7 above)
Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:1183]: Well, this is a mutating function... my guess would be that the fetched object gets injected into `osdCleanupTemplate`, so then we have to modify whatever is set to make sure it is reconciled properly. @xenolinux is that correct?

Comment (xenolinux) [reply pkg/controller/storagecluster/reconcile.go:1183]: Yes. That's right

File: pkg/controller/storagecluster/reconcile.go (hunk 8 above)
Comment (xenolinux) [reply pkg/controller/storagecluster/reconcile.go:1187]: ack

File: pkg/controller/storagecluster/reconcile.go [hunk 11]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
...
+	}
+	_, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, osdCleanUpTemplate, func() error {
+		osdCleanUpTemplate.Objects = []runtime.RawExtension{
+			{
+				Object: newCleanupJob(sc),
+			},
+		}
+		osdCleanUpTemplate.Parameters = []openshiftv1.Parameter{
+			{
+				Name:     "ROOK_IMAGE",
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1176]: We should be able to use the same image that is already being used for the rook operator. Can we lookup the rook operator image instead of requiring it as a parameter? The only input we should need from the admin is the osd id. The reconcile loop would also update the template whenever there is an upgrade to a new image.

File: pkg/controller/storagecluster/reconcile.go [hunk 12]
```diff
@@ -1152,3 +1156,129 @@ func getMonCount(nodeCount int) int {
...
+							Name:  "script",
+							Image: "${ROOK_IMAGE}",
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									MountPath: "/etc/ceph",
+									Name:      "ceph-config",
+									ReadOnly:  true,
+								},
+							},
+							Command: []string{"/bin/bash", "-c", "# ensure the osd is marked out\nceph osd out osd.3\n# purge the osd\nceph osd purge osd.3\n"},
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1245]: We need to replace the failed id `3` with the param. This is probably leftover from a test.

Comment (travisn) [reply pkg/controller/storagecluster/reconcile.go:1245]: For code readability I'd suggest using a golang multiline string

File: pkg/controller/storagecluster/reconcile.go (hunk 4 above)
Comment (ashishranjan738) [reply pkg/controller/storagecluster/reconcile.go:1160]: @xenolinux you need to wrap this around a user-friendly method

Comment (ashishranjan738) [reply pkg/controller/storagecluster/reconcile.go:1160]: reframe this statement as the implementation has been changed since it started

File: pkg/controller/storagecluster/reconcile.go (hunk 6 above)
Comment (ashishranjan738) [reply pkg/controller/storagecluster/reconcile.go:1168]: Umm I would still prefer this createOrUpdate method as this has been a recommended way of creating/updating the resources. https://godoc.org/sigs.k8s.io/controller-runtime/pkg/controller/controllerutil#CreateOrUpdate

File: pkg/controller/storagecluster/reconcile.go (hunk 7 above)
Comment (ashishranjan738) [reply pkg/controller/storagecluster/reconcile.go:1183]: yup can be done

File: pkg/controller/storagecluster/reconcile.go (hunk 11 above)
Comment (ashishranjan738) [reply pkg/controller/storagecluster/reconcile.go:1176]: right can be done

File: pkg/controller/storagecluster/reconcile.go (hunk 12 above)
Comment (ashishranjan738) [reply pkg/controller/storagecluster/reconcile.go:1245]: yup changed to string literal

File: pkg/controller/storagecluster/reconcile.go [hunk 13]
```diff
@@ -1152,3 +1156,134 @@ func getMonCount(nodeCount int) int {
...
+		osdCleanUpTemplate.Objects = []runtime.RawExtension{
+			{
+				Object: newCleanupJob(sc),
+			},
+		}
+		osdCleanUpTemplate.Parameters = []openshiftv1.Parameter{
+			{
+				Name:     "ROOK_IMAGE",
+				Required: true,
+				Value:    os.Getenv("ROOK_CEPH_IMAGE"),
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1178]: - This is an env var to the ocs operator? 
- Instead of the parameter, let's just set the image directly in the template spec. For example, on lines 1237 and 1256 just set it directly instead of setting it in the param here.

File: pkg/controller/storagecluster/reconcile.go [hunk 14]
```diff
@@ -1152,3 +1156,134 @@ func getMonCount(nodeCount int) int {
...
+							VolumeMounts: []corev1.VolumeMount{
+								{
+									MountPath: "/etc/ceph",
+									Name:      "ceph-config",
+									ReadOnly:  true,
+								},
+							},
+							Command: []string{"/bin/bash", "-c",
+								`# ensure the osd is marked out
+								 ceph osd out osd.FAILED_OSD_ID
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1247]: Do we need `${FAILED_OSD_ID}` similar to other places in the template, or why don't we need the brackets here?

File: pkg/controller/storagecluster/reconcile.go [hunk 15]
```diff
@@ -1152,3 +1156,134 @@ func getMonCount(nodeCount int) int {
...
+								 # purge the osd
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1248]: Note that the script will have lots of spaces before these lines. It shouldn't hurt the script, but it would be better if we could avoid that whitespace.

Comment (xenolinux) [reply pkg/controller/storagecluster/reconcile.go:1248]: ack

File: pkg/controller/storagecluster/reconcile.go (hunk 13 above)
Comment (xenolinux) [reply pkg/controller/storagecluster/reconcile.go:1178]: yes.

File: pkg/controller/storagecluster/reconcile.go [hunk 16]
```diff
@@ -1152,3 +1156,128 @@ func getMonCount(nodeCount int) int {
 
 	return count
 }
+
+// ensureJobTemplates ensures if the osd removal job template exists
+func (r *ReconcileStorageCluster) ensureJobTemplates(sc *ocsv1.StorageCluster, reqLogger logr.Logger) error {
+	osdCleanUpTemplate := &openshiftv1.Template{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sc.Name + "-osd-removal",
```
Comment (travisn) [inline pkg/controller/storagecluster/reconcile.go:1164]: Will the template be called `openshift-storage-osd-removal`? Or what is `sc.Name`? How about if we just use a constant for the template name? Then the documentation can use the well-known template name.

Comment (jarrpa) [reply pkg/controller/storagecluster/reconcile.go:1164]: This makes sense, it would simplify the documentation. I propose something simple like `ocs-osd-removal`.

File: pkg/controller/storagecluster/reconcile.go [hunk 17]
```diff
@@ -1152,3 +1156,128 @@ func getMonCount(nodeCount int) int {
...
+		Spec: batchv1.JobSpec{
+			Template: corev1.PodTemplateSpec{
+				Spec: corev1.PodSpec{
+					Volumes: []corev1.Volume{
+						{
+							Name: "mon-endpoint-volume",
+							VolumeSource: corev1.VolumeSource{
+								ConfigMap: &corev1.ConfigMapVolumeSource{
+									LocalObjectReference: corev1.LocalObjectReference{
+										Name: "mon-endpoint-volume",
```
Comment (ashishranjan738) [inline pkg/controller/storagecluster/reconcile.go:1216]: this name should be `rook-ceph-mon-endpoints`

**Comment by travisn** [comment]
> this name should be rook-ceph-mon-endpoints

@xenolinux @ashishranjan738 With all the changes could you also confirm what testing has been done? For example, that the template is created by the OCS operator and the job can be run successfully with the `oc process` command. I'm assuming this issue with the mon endpoints name was found through testing, just wanted to confirm once it's all working end to end. Thanks!

**Comment by jarrpa** [comment]
@xenolinux Okay, I found the problem I was having: you swashed the wrong commits. Currently, there are three commits:

1. 8ee281f8 -    5 hours ago Changed the name — Original changes
2. 19e88cc5 -    5 hours ago Adds vendor files — vendor files
3. 38963f46 -    5 hours ago Adds reconcile logic for removing osd — CSV and PR edits

At this point, I propose squashing commit 3 into commit 1 as a "fix" (e.g. without the commit message) that way we get all the code and CSV changes in one commit but still leave the vendor files separate. If you feel like it you can split the CSV changes into their own commit, but I'm willing to forego that in the interest of time.

**Comment by ashishranjan738** [comment]
> > this name should be rook-ceph-mon-endpoints
> 
> @xenolinux @ashishranjan738 With all the changes could you also confirm what testing has been done? For example, that the template is created by the OCS operator and the job can be run successfully with the `oc process` command. I'm assuming this issue with the mon endpoints name was found through testing, just wanted to confirm once it's all working end to end. Thanks!

umm I tested it and seems like its working fine
